components/script/serviceworker_manager.rs

```diff
@@ -108,76 +188,194 @@ impl ServiceWorkerManager {
-                true
-            },
-            ServiceWorkerMsg::Timeout(scope) => {
-                if self.active_workers.contains_key(&scope) {
-                    let _ = self.active_workers.remove(&scope);
-                } else {
-                    warn!("ServiceWorker for {:?} is not active", scope);
-                }
-                true
-            },
+            ServiceWorkerMsg::Timeout(scope) => {},
             ServiceWorkerMsg::ForwardDOMMessage(msg, scope_url) => {
-                if self.active_workers.contains_key(&scope_url) {
-                    if let Some(ref sender) = self.active_workers.get(&scope_url) {
-                        self.forward_message(msg, &sender);
-                    }
-                } else {
-                    if let Some(ref sender) = self.wakeup_serviceworker(scope_url) {
-                        self.forward_message(msg, &sender);
+                if let Some(registration) = self.registrations.get_mut(&scope_url) {
+                    if let Some(ref worker) = registration.active_worker {
+                        worker.forward_dom_message(msg);
                     }
                 }
-                true
             },
-            ServiceWorkerMsg::Exit => false,
+            ServiceWorkerMsg::ScheduleJob(job) => match job.job_type {
+                JobType::Register => {
+                    self.handle_register_job(job);
+                },
+                JobType::Update => {
+                    self.handle_update_job(job);
+                },
+                JobType::Unregister => {
+                    // TODO: https://w3c.github.io/ServiceWorker/#unregister-algorithm
+                },
+            },
+            ServiceWorkerMsg::Exit => return false,
         }
+        true
     }
 
-    fn handle_message_from_resource(&mut self, mediator: CustomResponseMediator) -> bool {
-        if serviceworker_enabled() {
-            if let Some(scope) = self.get_matching_scope(&mediator.load_url) {
-                if self.active_workers.contains_key(&scope) {
-                    if let Some(sender) = self.active_workers.get(&scope) {
-                        let _ = sender.send(ServiceWorkerScriptMsg::Response(mediator));
-                    }
-                } else {
-                    if let Some(sender) = self.wakeup_serviceworker(scope) {
-                        let _ = sender.send(ServiceWorkerScriptMsg::Response(mediator));
-                    }
-                }
-            } else {
-                let _ = mediator.response_chan.send(None);
+    /// <https://w3c.github.io/ServiceWorker/#register-algorithm>
+    fn handle_register_job(&mut self, mut job: Job) {
+        if !job.script_url.is_origin_trustworthy() {
+            // Step 1.1
+            let _ = job
+                .client
+                .send(JobResult::RejectPromise(JobError::SecurityError));
+            return;
+        } else if job.script_url.origin() != job.referrer.origin() ||
```

Comment:
Due to different from checking origin trustworthy and different steps, maybe it's good to have it as another standalone `if` expression?

```rust
if !job.script_url.is_origin_trustworthy() {
  ...
}

if job.script_url.origin() != job.referrer.origin() ||
  job.scope_url.origin() != job.referrer.origin()
{
  ...
}
```